cpp/splinepy/utils/arrays.hpp

```diff
@@ -0,0 +1,283 @@
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <algorithm>
+#include <numeric>
+
+namespace splinepy::utils {
+
+using namespace splinelib::sources;  
+
+/// Elementwise subtraction
+template<typename T, typename NamedType, std::size_t dim>
+inline void FirstMinusSecondEqualsThird(
+    const std::array<NamedType, dim>& first,
+    const T* second, /* c array */
+    std::array<T, dim>& third) {
+  for (size_t i{}; i < dim; ++i) {
+    // following line should raise error during compile time
+    // if T != NameType::Type_
+    third[i] = static_cast<T>(first[i]) - second[i];
+  }
+}
+
+/// Elementwise mean
+template<typename T, std::size_t dim>
+inline std::array<T, dim> Mean(const std::array<T, dim>& arr1,
+                               const std::array<T, dim>& arr2) {
+  std::array<T, dim> out;
+  for (int i{0}; i < dim; ++i) {
+    out[i] = (arr1[i] + arr2[i]) * .5;
+  }
+  return out;
+}
+
+/// Elementwise mean overload, you can specified return value_type.
+/// Useful for returning NamedType.
+template<typename ReturnT, typename T, std::size_t dim>
+inline std::array<ReturnT, dim> Mean(const std::array<T, dim>& arr1,
+                                     const std::array<T, dim>& arr2) {
+  std::array<ReturnT, dim> out;
+  for (int i{0}; i < dim; ++i) {
+    out[i] = ReturnT{(arr1[i] + arr2[i]) * .5};
+  }
+  return out;
+}
+
+/// Dot product
+template<typename T1, typename T2, std::size_t dim>
+inline T1 Dot(const std::array<T1, dim>& arr1,
+              const std::array<T2, dim>& arr2) {
+  T1 dotted{}; /* default value should be 0. */
+  for (int i{0}; i < dim; ++i) {
+    dotted += arr1[i] * static_cast<T1>(arr2[i]);
+  }
+  return dotted;
+}
+
+/// AAt
+template<typename T, std::size_t dim1, std::size_t dim2>
+inline std::array<std::array<T, dim1>, dim1> AAt(
+    const std::array<std::array<T, dim2>, dim1>& arr1) {
+
+  std::array<std::array<T, dim1>, dim1> out;
+
+  for (int i{}; i < dim1; ++i) {
+    for (int j{i}; j < dim2; ++j) {
+      out[i][j] = Dot(arr1[i], arr1[j]);
+      if (i != j) out[j][i] = out[i][j]; // fill symmetric part
+    }
+  }
+
+  return out;
+}
+
+/// elementwise inplace addition
+template<typename T1, typename T2, std::size_t dim>
+inline void AddSecondToFirst(std::array<T1, dim>& arr1,
+                             const std::array<T2, dim>& arr2) {
+  for (int i{0}; i < dim; i++) {
+    arr1[i] += T1{arr2[i]};
+  }
+}
+
+/// elementwise inplace addition with a given factor.
+template<typename T1, typename T2, std::size_t dim>
+inline void AddSecondTimesThirdToFirst(std::array<T1, dim>& arr1,
```

Comment:
This is inconsistent. In the first function you take the third as return argument. You should use the same outline here!